### PR TITLE
[BACKEND] Update yt-dlp, move to pyproject.toml

### LINKED ISSUE #961: Update dependencies
Making the dependencies hard on one version makes it much more difficult to package especially on distros like NixOS where in order to package it, the dependencies each need to be packaged as well in the case of this project. Perhaps it might be a good idea to instead of depending on specific versions, depend on their major (or major+minor) version numbers?

I also think moving to a `pyproject.toml` file wouldn't be a bad idea either, since this will allow the removal of the setup.cfg file.

Of course, I'm not just saying this and putting the burden on you, I'm willing to do this work. I am more gauging the interest in this.

I've already tested at least updating the PyYAML dependency and removing the `argparse` dependency (since it comes in the python stdlib), and it worked flawlessly. All drop-in.

### PR DESCRIPTION
**Pull request by Noodlez1232**
This also includes some dependency updates. I would really appreciate a review of them.

I don't know much about how the versioning in python works (I am not really a Python developer), but I know there's instead [a type of versioning](https://packaging.python.org/en/latest/specifications/version-specifiers/#id5) that works more like this: `yt-dlp>=2023.05.03` where any version newer than that version is compatible. I was wondering if maybe it should move to this? It might break the project in the future, but it'd be a loud break, meaning it'd be easy to catch and fix. [There's also a type called "Compatible release"](https://packaging.python.org/en/latest/specifications/version-specifiers/#compatible-release) that works like this: `sphinx~=7.2.6` that might be useful for some of the dependencies that are more important to pin.

Fixes #961

### CODE REVIEW

**Comment by Noodlez1232** [comment]
To clarify, I did *not* test or do anything that has to do with deps yet, I'm waiting for you to take a look and see what you think, and then I'll continue with that part. It's a bit difficult to test with deps on NixOS, so I'm waiting for that.

File: pyproject.toml [hunk 1]
```diff
@@ -1,3 +1,69 @@
...
+classifiers = [
+    "Topic :: Multimedia :: Sound/Audio",
+    "Topic :: Multimedia :: Video",
+    "License :: Public Domain",
+    "Environment :: Console",
+    "Programming Language :: Python :: 3.10",
+    "Programming Language :: Python :: 3.11",
+]
+dependencies = [
+    "yt-dlp==2024.*",
```
Comment (jmbannon) [inline pyproject.toml:18]: We probably want to keep this one as exact since it's susceptible to break ytdl-sub

**Comment by jmbannon** [comment]
This looks great! With yt-dlp I've been meaning to set up auto-deps which creates a PR automatically with the new version. The other deps should be safe to allow minor versions

**Comment by Noodlez1232** [comment]
Looking through the test failures, it seems there may be some sort of dependency failure between `pytest-rerunfailures` and `ytdl-sub` since they require different versions of the same package. I'll look into this as well. For now, I'll keep messing with it and see what I can do in terms of dependency resolution and all that.
I'll make sure I tighten up the dependency on `yt-dlp` as well. I'll keep it pinned.

**Comment by Noodlez1232** [comment]
Btw looking into the problem of the PR, it seems that [dependencies will be able to be dynamically generated soon (using setuptools)](https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html#dynamic-metadata). It's in beta right now. It'll dynamically pull that info from a `requirements.txt` type file which'd be easy to `sed` in order to auto-generate a yt-dlp update pull using CI. Just make it always line 0 or something and then you'd be able to just generate it on-the-fly. I don't know how it interacts with the existing dependencies specifier in `[project]`, whether it just adds to it, has precedence, or overwrites those values completely.
Of course, it's in beta, so probably not important right now, but maybe one day.

**Comment by Noodlez1232** [comment]
I'm hoping that worked... I had some trouble getting the environment up, so idk if it worked.

**Comment by jmbannon** [comment]
@Noodlez1232  Going to keep the lint deps stationary to avoid having this PR alter anything there. Everything else looks good! Once the build tests pass, I will merge

**Pull request merged**